core/include/gnuradio-4.0/Scheduler.hpp

```diff
@@ -20,212 +24,283 @@ using namespace std::string_literals;
                 }
             }
-        });
-        _state = lifecycle::State::STOPPED;
-    }
-
-    void
-    pause() {
-        if (!canPause()) {
-            return;
         }
-        requestPause();
-        waitDone();
-        this->_graph.forEachBlock([this](auto &block) {
-            if (auto e = block.changeState(lifecycle::State::REQUESTED_PAUSE); !e) {
-                auto                       &port  = msgOut;
-                const lifecycle::ErrorType &error = e.error();
-                this->emitMessage(port, { { key::Kind, kind::Error }, { key::ErrorInfo, error.message }, { key::Location, error.srcLoc() } });
-            }
-        });
-        _state = lifecycle::State::PAUSED;
+        return {};
     }
 
     void
     waitDone() {
-        using enum lifecycle::State;
         [[maybe_unused]] const auto pe = _profiler_handler.startCompleteEvent("scheduler_base.waitDone");
-        for (auto running = _running_jobs.load(); running > 0ul; running = _running_jobs.load()) {
-            this->processScheduledMessages();
-            std::this_thread::sleep_for(std::chrono::milliseconds(10));
+        for (auto running = _running_jobs.load(); this->state() == lifecycle::REQUESTED_PAUSE || this->state() == lifecycle::PAUSED || running > 0ul; running = _running_jobs.load()) {
+            std::this_thread::sleep_for(kMessagePollInterval);
+            processScheduledMessages();
         }
-        this->processScheduledMessages();
+        processScheduledMessages();
+    }
+
+    [[nodiscard]] const std::vector<std::vector<BlockModel *>> &
+    jobs() const {
+        return _job_lists;
     }
 
+protected:
     void
-    requestStop() {
-        _stop_requested = true;
-        _state          = lifecycle::State::REQUESTED_STOP;
+    emitMessage(auto &port, Message message) { // TODO: to be replaced if scheduler derives from Block<T>
+        message[key::Sender] = unique_name;
+        port.streamWriter().publish([&](auto &out) { out[0] = std::move(message); }, 1);
     }
 
     void
-    requestPause() {
-        _stop_requested = true;
-        _state          = lifecycle::State::REQUESTED_PAUSE;
+    emitError(auto &port, const lifecycle::ErrorType &error) {
+        emitMessage(port, { { key::Kind, kind::Error }, { key::ErrorInfo, error.message }, { key::Location, error.srcLoc() } });
     }
 
     void
-    connectBlockMessagePorts() {
-        _graph.forEachBlock([this](auto &block) {
-            if (ConnectionResult::SUCCESS != _toChildMessagePort.connect(*block.msgIn)) {
-                this->emitMessage(msgOut, { { key::Kind, kind::Error },
-                                      { key::ErrorInfo, fmt::format("Failed to connect scheduler input message port to child '{}'", block.uniqueName()) },
-                                      { key::Location, fmt::format("{}", std::source_location::current()) } });
+    setBlocksState(lifecycle::State state) {
```

Comment:
I spent quite some time on this and I think -- similiar to the emitError -- this function should not be explicitely defined but inlined for the time being. Similarly for the ` _graph.forEachBlock(...)`.

The rationale is simplicity, lean and clean design with least surprises or custom API that users and/or maintainers need to learn. 

Both functions combined are basic several onion layers of standard C++ functions that by renaming them obfuscates a bit what happens under the hood. In this case the `_graph.forEachBlock` is basically -- for example/illustration -- a
```cpp
std::ranges::for_each(_blocks, [f](const auto &block_ptr) { std::invoke(f, *block_ptr.get()); });
```
The emitError is basically a 
```cpp
 emitMessage(port, { { key::Kind, kind::Error }, { key::ErrorInfo, error.message }, { key::Location, error.srcLoc() } });
```
which hides the keys/fields that are being set into the method name (and creates an issue that the source location isn't the calling side but the location where emitError is being defined).

I think creating new public API here and at stage -- I know this is/was well meaning -- is PMO and we should introduce only new API when it's actually needed and all the other uses cases are clear. Especially w.r.t. the `emitError` case, it's mentioned already above that the `emitMessage` should be irradicated in favour of the scheduler being based on the `Block<T>` API  where these type of definitions and/or parallel CRTP (similar to the `lifecycle::StateMachine<T>`) are better placed.

Until then I think it's better to keep these definitions (especially there are only a few of them) explicited and unwrapped to the minimum API that is presently defined.

During the next iterations/PR, we should also remove `forEachBlock` from `Graph` since this is just a very thin wrapper that adds more costs in terms of cognitive complexity than benefit (at least at the moment).

Reply:
Done, removed both emitError() and setBlocksState() abstractions.